OpenRA.Game/Primitives/PriorityQueue.cs

```diff
@@ -22,98 +22,145 @@ public interface IPriorityQueue<T>
 		T Pop();
 	}
 
-	public class PriorityQueue<T> : IPriorityQueue<T>
+	/// <summary>
+	/// Represents a collection of items that have a priority.
+	/// On pop, the item with the lowest priority value is removed.
+	/// </summary>
+	public sealed class PriorityQueue<T, TComparer> : IPriorityQueue<T> where TComparer : struct, IComparer<T>
 	{
-		readonly List<T[]> items;
-		readonly IComparer<T> comparer;
-		int level, index;
-
-		public PriorityQueue()
-			: this(Comparer<T>.Default) { }
-
-		public PriorityQueue(IComparer<T> comparer)
+		/// <summary>
+		/// Compares two items to determine their priority.
+		/// PERF: Using a struct allows the calls to be devirtualized.
+		/// </summary>
+		readonly TComparer comparer;
+
+		/// <summary>
+		/// A min-heap storing the items. An array divided into sub arrays called levels.
+		/// At each level the size of a level array doubles.
+		/// Elements at deeper levels always have higher priority than elements nearer to the root.
+		/// </summary>
+		T[] items;
+
+		/// <summary>
+		/// Index of deepest level.
+		/// </summary>
+		int level;
+
+		/// <summary>
+		/// Number of elements in the deepest level.
+		/// </summary>
+		int index;
+
+		public PriorityQueue(TComparer comparer)
 		{
-			items = new List<T[]> { new T[1] };
 			this.comparer = comparer;
+			items = new T[1];
 		}
 
 		public void Add(T item)
 		{
 			var addLevel = level;
 			var addIndex = index;
 
-			while (addLevel >= 1 && comparer.Compare(Above(addLevel, addIndex), item) > 0)
+			while (addLevel >= 1)
 			{
-				items[addLevel][addIndex] = Above(addLevel, addIndex);
-				--addLevel;
-				addIndex >>= 1;
+				var above = items[AboveIndex(addLevel, addIndex)];
+				if (comparer.Compare(above, item) > 0)
+				{
+					items[Index(addLevel, addIndex)] = above;
+					--addLevel;
+					addIndex >>= 1;
+				}
+				else
+					break;
 			}
 
-			items[addLevel][addIndex] = item;
+			items[Index(addLevel, addIndex)] = item;
 
 			if (++index >= 1 << level)
 			{
 				index = 0;
-				if (items.Count <= ++level)
-					items.Add(new T[1 << level]);
+				var levelWidth = 1 << ++level;
+				if (levelWidth + levelWidth - 1 >= items.Length)
+					GrowCapacity();
 			}
 		}
 
+		void GrowCapacity()
+		{
+			var newItems = new T[(1 << level) * 2];
+			Array.Copy(items, newItems, items.Length);
+			items = newItems;
```

Comment:
Isn't repetitive allocations of new arrays _(and discarding the old ones)_ counter-productive?

Reply:
it shouldn't happen often (after the first few layers) so most likely this is better for perf than using a list

Reply:
We could instead of creating the GrowCapacity function call the Array.Resize. Functionally, it's the exact same, but perhaps dotnet has some extra optimisations in it that aren't present here

Reply:
Updated to use `Array.Resize`.

I don't think the resizing here should be particularly detrimental, but if we prove it to be an issue at some point we can teach calling code to pass an initial capacity in the same way you would set the capacity of other collections like `List`, `Dictionary`, etc in advance when it matters.